pkg/watch/watch.go

```diff
@@ -202,60 +216,227 @@ func addRecursiveWatch(watcher *fsnotify.Watcher, rootPath string, path string,
 	if err != nil {
-		return fmt.Errorf("error watching source path %s: %v", parameters.Path, err)
+		return nil, fmt.Errorf("error watching source path %s: %v", path, err)
 	}
-
-	printInfoMessage(out, parameters.Path)
-
-	return eventWatcher(ctx, watcher, parameters, out, evaluateFileChanges, processEvents, o.CleanupDevResources)
+	return watcher, nil
 }
 
 // eventWatcher loops till the context's Done channel indicates it to stop looping, at which point it performs cleanup.
 // While looping, it listens for filesystem events and processes these events using the WatchParameters to push to the remote pod.
 // It outputs any logs to the out io Writer
-func eventWatcher(ctx context.Context, watcher *fsnotify.Watcher, parameters WatchParameters, out io.Writer, evaluateChangesHandler evaluateChangesFunc, processEventsHandler processEventsFunc, cleanupHandler cleanupFunc) error {
+func (o *WatchClient) eventWatcher(ctx context.Context, sourcesWatcher *fsnotify.Watcher, deploymentWatcher watch.Interface, devfileWatcher *fsnotify.Watcher, podWatcher watch.Interface, eventsWatcher watch.Interface, parameters WatchParameters, out io.Writer, evaluateChangesHandler evaluateChangesFunc, processEventsHandler processEventsFunc, cleanupHandler cleanupFunc, componentStatus ComponentStatus) error {
+
+	expBackoff := NewExpBackoff()
+
 	var events []fsnotify.Event
 
-	// timer helps collect multiple events that happen in a quick succession. We start with 1ms as we don't care much
-	// at this point. In the select block, however, every time we receive an event, we reset the timer to watch for
+	// sourcesTimer helps collect multiple events that happen in a quick succession. We start with 1ms as we don't care much
+	// at this point. In the select block, however, every time we receive an event, we reset the sourcesTimer to watch for
 	// 100ms since receiving that event. This is done because a single filesystem event by the user triggers multiple
 	// events for fsnotify. It's a known-issue, but not really bug. For more info look at below issues:
 	//    - https://github.com/fsnotify/fsnotify/issues/122
 	//    - https://github.com/fsnotify/fsnotify/issues/344
-	timer := time.NewTimer(time.Millisecond)
-	<-timer.C
+	sourcesTimer := time.NewTimer(time.Millisecond)
+	<-sourcesTimer.C
+
+	devfileTimer := time.NewTimer(time.Millisecond)
+	<-devfileTimer.C
+
+	deployTimer := time.NewTimer(time.Millisecond)
+	<-deployTimer.C
+
+	retryTimer := time.NewTimer(time.Millisecond)
+	<-retryTimer.C
+
+	podsPhases := NewPodPhases()
 
 	for {
 		select {
-		case event := <-watcher.Events:
+		case event := <-sourcesWatcher.Events:
 			events = append(events, event)
 			// We are waiting for more events in this interval
-			timer.Reset(100 * time.Millisecond)
-		case <-timer.C:
+			sourcesTimer.Reset(100 * time.Millisecond)
+		case <-sourcesTimer.C:
 			// timer has fired
+			if !componentCanSyncFile(componentStatus.State) {
```

Comment:
No, the function is really made to tell if sync can be done. It appears that for the moment, it is testing if the component is ready, but it will most probably evolve in the future.